### PR TITLE
[Operator] Init NLL_LOSS

### PR DESCRIPTION
**Pull request by GwokHiujin**
A basic implementation of `NLL_LOSS` has been pushed.

Based on the performance testing results summarized earlier, we believe that using the `gather` operation would lead to a more efficient implementation (by observing the output results of latency, it seems this is also how torch does it), and we will push forward with this optimization.

### CODE REVIEW

**Review by tongxin — APPROVED** [review]
LGTM

File: benchmark/test_reduction_perf.py [hunk 1]
```diff
@@ -135,6 +141,13 @@ def cumsum_input_fn(shape, cur_dtype, device):
             FLOAT_DTYPES + INT_DTYPES,
             marks=pytest.mark.cumsum,
         ),
+        pytest.param(
+            "nll_loss",
+            torch.nn.NLLLoss,
```
Comment (tongxin) [inline benchmark/test_reduction_perf.py]: NLLLoss is a class. Can we use it as the reference function?

File: src/flag_gems/ops/nllloss.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    w_ptr,
+    w_tgt_ptr,
+    out_ptr,
+    ignore_index,
+    N,
+    C,
+    BLOCK_N: tl.constexpr,
+):
+    pid_n = tl.program_id(0)
+    offsets_n = pid_n * BLOCK_N + tl.arange(0, BLOCK_N)
```
Comment (tongxin) [inline src/flag_gems/ops/nllloss.py:24]: Check `offset_n` overflow.

File: src/flag_gems/ops/nllloss.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+    ignore_index,
+    N,
+    C,
+    D,
+    BLOCK_C: tl.constexpr,
+    BLOCK_D: tl.constexpr,
+):
+    pid_n = tl.program_id(0)
+    pid_d = tl.program_id(1)
+    offset_d = pid_d * BLOCK_D + tl.arange(0, BLOCK_D)
```
Comment (tongxin) [inline src/flag_gems/ops/nllloss.py]: Overflow check.

File: src/flag_gems/ops/nllloss.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    ignore_index,
+    total_weight,
+    N,
+    C,
+    BLOCK_N: tl.constexpr,
+    BLOCK_C: tl.constexpr,
+):
+    pid_n = tl.program_id(0)
+    pid_c = tl.program_id(1)
+    offsets_n = pid_n * BLOCK_N + tl.arange(0, BLOCK_N)
```
Comment (tongxin) [inline src/flag_gems/ops/nllloss.py:77]: Overflow check.

File: src/flag_gems/ops/nllloss.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+        axis = 0 if dim == 1 else 1
+        del shape[axis]
+
+        assert ((i >= 0 and i < C) for i in target), "Target is out of bounds"
+        assert list(target.shape) == shape, "Invalid target size"
+        assert inp.ndim >= 1, "Invalid input ndim"
+
+        if weight is None:
+            weight = torch.ones(
+                [
```
Comment (tongxin) [inline src/flag_gems/ops/nllloss.py]: Use tuple

File: src/flag_gems/ops/nllloss.py [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+        )
+        tl.store(inp_grad_ptrs, inp_grad.to(tl.float32), mask=(inp_mask & ignore_mask))
+
+
+class NLLLoss(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, inp, target, weight, reduction, ignore_index):
+        logging.debug("GEMS NLLLoss FWD")
+        shape = list(inp.shape)
+        dim = inp.ndim
```
Comment (tongxin) [inline src/flag_gems/ops/nllloss.py]: The input shape/layout appears to be pretty complex as shown in the pytorch doc. Shall we add some documentation here to help clarify the inputs?

File: src/flag_gems/ops/nllloss.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+    total_weight,
+    N,
+    C,
+    D,
+    BLOCK_C: tl.constexpr,
+    BLOCK_D: tl.constexpr,
+):
+    pid_n = tl.program_id(0)
+    pid_d = tl.program_id(1)
+    offset_d = pid_d * BLOCK_D + tl.arange(0, BLOCK_D)
```
Comment (tongxin) [inline src/flag_gems/ops/nllloss.py]: Overflow check.

File: src/flag_gems/ops/nllloss.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+        inp_mask = offset_c[:, None] < C and offset_d[None, :] < D
+        inp_mask = inp_mask and offset_c[:, None] == tgt
+        inp_grad = -1 * out_grad * w_tgt / total_weight
+        inp_grad_ptrs = (
+            inp_grad_ptr + pid_n * C * D + offset_c[:, None] * D + offset_d[None, :]
+        )
+        tl.store(inp_grad_ptrs, inp_grad.to(tl.float32), mask=(inp_mask & ignore_mask))
+
+
+class NLLLoss(torch.autograd.Function):
```
Comment (tongxin) [inline src/flag_gems/ops/nllloss.py]: This function is intended be used as substitute for `nll_loss` whereas NLLLoss is already taken as the nn module name. We should avoid the name confusion.

File: src/flag_gems/ops/nllloss.py [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+        ctx.D = D
+        ctx.ignore_index = ignore_index
+        ctx.total_weight = 1
+        ctx.shape = shape
+
+        # redution: 0-None, 1-mean, 2-sum
+        if reduction == 0:
+            res = out.to(inp.dtype)
+        elif reduction == 1:
+            ctx.total_weight = sum(w_tgt).item()
```
Comment (tongxin) [inline src/flag_gems/ops/nllloss.py]: Shall we also add `dim=` args to avoid confusion?

File: benchmark/test_reduction_perf.py (hunk 1 above)
Comment (GwokHiujin) [reply benchmark/test_reduction_perf.py]: Indeed. I've updated it to `torch.nn.functional.nll_loss`.

File: src/flag_gems/ops/nllloss.py (hunk 2 above)
Comment (GwokHiujin) [reply src/flag_gems/ops/nllloss.py:24]: Sorry, I didn't understand this suggestion. Shouldn't the subsequent *mask generation* be sufficient to handle potential overflow? Or are you suggesting we check if `offsets_n` could exceed Triton's maximum representable number? If so, I think many operators will need to incorporate this check too.

File: src/flag_gems/ops/nllloss.py (hunk 3 above)
Comment (GwokHiujin) [reply src/flag_gems/ops/nllloss.py]: ditto

File: src/flag_gems/ops/nllloss.py (hunk 4 above)
Comment (GwokHiujin) [reply src/flag_gems/ops/nllloss.py:77]: ditto

File: src/flag_gems/ops/nllloss.py (hunk 5 above)
Comment (GwokHiujin) [reply src/flag_gems/ops/nllloss.py]: Done.

File: src/flag_gems/ops/nllloss.py (hunk 8 above)
Comment (GwokHiujin) [reply src/flag_gems/ops/nllloss.py]: Indeed. I've upated the class name.

File: src/flag_gems/ops/nllloss.py (hunk 6 above)
Comment (GwokHiujin) [reply src/flag_gems/ops/nllloss.py]: I've tried to add some annotations here about nllloss's parameters. 

However, I think users who will use this function should also be quite clear about its principles; the part that looks complicated is just a high-dimensional input 😄

Comment (tongxin) [reply src/flag_gems/ops/nllloss.py]: Thanks for adding the doc. It always helps!

**Review by tongxin — DISMISSED** [review]
LGTM

**Comment by StrongSpoon** [comment]
could you provide the performance data?

**Comment by GwokHiujin** [comment]
> could you provide the performance data?

A general result is as below: 

```
Operator: nll_loss  Performance Test (dtype=torch.float16, mode=cuda, level=comprehensive)
Size         Torch Latency (ms)    Gems Latency (ms)         Gems Speedup         Size Detail
------------------------------------------------------------------------------------------
SUCCESS               0.008192            0.430080               0.019          [torch.Size([64, 64]), torch.Size([64])]
SUCCESS               0.014336            0.626688               0.023          [torch.Size([256, 256]), torch.Size([256])]
SUCCESS               0.039936            1.439744               0.028          [torch.Size([1024, 1024]), torch.Size([1024])]
SUCCESS               0.146432            4.753408               0.031          [torch.Size([4096, 4096]), torch.Size([4096])]
SUCCESS               0.039936            1.436672               0.028          [torch.Size([1024, 65536]), torch.Size([1024])]
SUCCESS               0.037888            1.439744               0.026          [torch.Size([1024, 1]), torch.Size([1024])]
SUCCESS               0.034816            1.430528               0.024          [torch.Size([1024, 16]), torch.Size([1024])]
SUCCESS               0.043008            1.440768               0.030          [torch.Size([1024, 256]), torch.Size([1024])]
SUCCESS               0.039936            1.432576               0.028          [torch.Size([1024, 4096]), torch.Size([1024])]


Operator: nll_loss  Performance Test (dtype=torch.float32, mode=cuda, level=comprehensive)
Size         Torch Latency (ms)    Gems Latency (ms)         Gems Speedup         Size Detail
------------------------------------------------------------------------------------------
SUCCESS               0.008192            0.413696               0.020          [torch.Size([64, 64]), torch.Size([64])]
SUCCESS               0.014336            0.601088               0.024          [torch.Size([256, 256]), torch.Size([256])]
SUCCESS               0.039936            1.414144               0.028          [torch.Size([1024, 1024]), torch.Size([1024])]
SUCCESS               0.146432            4.697088               0.031          [torch.Size([4096, 4096]), torch.Size([4096])]
SUCCESS               0.038912            1.420288               0.027          [torch.Size([1024, 65536]), torch.Size([1024])]
SUCCESS               0.036864            1.413120               0.026          [torch.Size([1024, 1]), torch.Size([1024])]
SUCCESS               0.034816            1.412096               0.025          [torch.Size([1024, 16]), torch.Size([1024])]
SUCCESS               0.043008            1.417216               0.030          [torch.Size([1024, 256]), torch.Size([1024])]
SUCCESS               0.038912            1.411072               0.028          [torch.Size([1024, 4096]), torch.Size([1024])]


Operator: nll_loss  Performance Test (dtype=torch.bfloat16, mode=cuda, level=comprehensive)
Size         Torch Latency (ms)    Gems Latency (ms)         Gems Speedup         Size Detail
------------------------------------------------------------------------------------------
SUCCESS               0.008192            0.431104               0.019          [torch.Size([64, 64]), torch.Size([64])]
SUCCESS               0.014336            0.621568               0.023          [torch.Size([256, 256]), torch.Size([256])]
SUCCESS               0.039936            1.448960               0.028          [torch.Size([1024, 1024]), torch.Size([1024])]
SUCCESS               0.146432            4.698112               0.031          [torch.Size([4096, 4096]), torch.Size([4096])]
SUCCESS               0.039936            1.448960               0.028          [torch.Size([1024, 65536]), torch.Size([1024])]
SUCCESS               0.037888            1.452032               0.026          [torch.Size([1024, 1]), torch.Size([1024])]
SUCCESS               0.034816            1.444864               0.024          [torch.Size([1024, 16]), torch.Size([1024])]
SUCCESS               0.043008            1.442816               0.030          [torch.Size([1024, 256]), torch.Size([1024])]
SUCCESS               0.038912            1.446912               0.027          [torch.Size([1024, 4096]), torch.Size([1024])]
```

It is somewhat poor now. As mentioned before, we may use `gather` later for optimization.

File: src/flag_gems/ops/nllloss.py [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+                    inp, tgt, w, w_tgt, out, ignore_index, N, C, D
+                )
+        else:
+            grid = lambda meta: (triton.cdiv(N, meta["BLOCK_N"]),)
+            with torch.cuda.device(inp.device):
+                nll_loss_2d_fwd_kernel[grid](
+                    inp, tgt, w, w_tgt, out, ignore_index, N, C
+                )
+
+        ctx.save_for_backward(inp, tgt, w)
```
Comment (StrongSpoon) [inline src/flag_gems/ops/nllloss.py]: only saving tensors and variables when input requires gradient might decrease the cost.

File: src/flag_gems/ops/nllloss.py [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+        ctx.ignore_index = ignore_index
+        ctx.total_weight = 1
+        ctx.shape = shape
+
+        # redution: 0-None, 1-mean, 2-sum
+        if reduction == 0:
+            res = out.to(inp.dtype)
+        elif reduction == 1:
+            ctx.total_weight = sum(w_tgt).item()
+            res = sum(out).to(inp.dtype) / ctx.total_weight
```
Comment (StrongSpoon) [inline src/flag_gems/ops/nllloss.py]: suggest fusing sum into forward kernel, referencing to cross_entropy_loss.

**Comment by StrongSpoon** [comment]
performance after optimization:
截屏2025-01-06 14 26 09

File: src/flag_gems/ops/nllloss.py [hunk 12]
```diff
@@ -5,12 +5,12 @@
...
 
 from ..utils import libentry
-from .sum import sum
+from .cross_entropy_loss import sum_and_scale
 
 
 @libentry()
 @triton.autotune(
-    configs=[triton.Config({"BLOCK_N": n}, num_warps=4) for n in [256, 512, 1024]],
+    configs=[triton.Config({"BLOCK_N": n}, num_warps=4) for n in [1, 16, 256]],
```
Comment (tongxin) [inline src/flag_gems/ops/nllloss.py]: Why BLOCK_N varies so much?

File: src/flag_gems/ops/nllloss.py [hunk 13]
```diff
@@ -93,11 +92,7 @@ def nll_loss_2d_bwd_kernel(
 
 @libentry()
 @triton.autotune(
-    configs=[
-        triton.Config({"BLOCK_C": c, "BLOCK_D": d}, num_warps=4)
-        for c in [256, 512, 1024]
-        for d in [1, 4, 16]
-    ],
+    configs=[triton.Config({"BLOCK_D": d}, num_warps=4) for d in [1, 4, 16]],
```
Comment (tongxin) [inline src/flag_gems/ops/nllloss.py]: Considering this as gather/scatter like kernel, BLOCK = 128 should be good enough.

File: src/flag_gems/ops/nllloss.py (hunk 12 above)
Comment (tongxin) [reply src/flag_gems/ops/nllloss.py]: I think BLOCK = 128 is good pick.

Comment (StrongSpoon) [reply src/flag_gems/ops/nllloss.py]: what about [1, 4, 32, 128]?

**Review by tongxin — DISMISSED** [review]
LGTM

**Review by tongxin — DISMISSED** [review]

**Comment by StrongSpoon** [comment]
the latest performance:
截屏2025-01-15 09 43 19

File: src/flag_gems/ops/nllloss.py [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+        total_wgt = tl.sum(wgt_tgt)
+        tl.atomic_add(out_ptr, total_out, sem="relaxed")  # output
+        tl.atomic_add(out_ptr + 1, total_wgt, sem="relaxed")  # weight
+        tl.atomic_add(out_ptr + 2, 1, sem="release")  # counter
+        counter = tl.load(out_ptr + 2)
+        total_out = tl.load(out_ptr)
+        total_wgt = tl.load(out_ptr + 1)
+        tl.store(
+            out_ptr + 3, total_out / total_wgt, mask=(counter == tl.num_programs(0))
+        )
```
Comment (tongxin) [inline src/flag_gems/ops/nllloss.py]: I think it's safer to use a stronger memory order for the counter update in line 49. And then we're ensured to have only one CTA to do the rest.

Comment (StrongSpoon) [reply src/flag_gems/ops/nllloss.py]: release is enough here. it's safe even if there are more than one CTAs satisfying the condition.

**Review by tongxin — APPROVED** [review]
Looks good now.
